File: cpp_lib/convert_functions.cpp
#include "convert_functions.h"
#include "regexes.h"
#include <iostream>
#include <regex>
#include <unordered_map>
#include <vector>
#include <cassert>

// Function definitions

std::string easy_RLE_to_txt(const std::string& erttRLE) {
    std::smatch match;
    if (std::regex_search(erttRLE, match, RLE_regex)) {
        int x = std::stoi(match[3]);
        int y = std::stoi(match[4]);
        std::string rle = match[5];
        std::string rle_no_ws;
        for (char c : rle) {
            if (c != '\n' && c != '\t' && c != ' ') {
                rle_no_ws += c;
            }
        }
        return advanced_RLE_to_txt(x, y, rle_no_ws);
    }
    return "";
}

std::string advanced_RLE_to_txt(int arttXBound, int arttYBound, const std::string& arttRLE) {
    std::string arttOutput;
    std::string arttInt;
    int arttCharsInLine = 0;

    for (char c : arttRLE) {
        if (c == '\n' || c == '\t' || c == ' ') {
            continue;
        }
        if (std::isdigit(c)) {
            arttInt += c;
        } else if (c == '$') {
            if (arttInt.empty()) arttInt = "1";
            for (int i = 0; i < std::stoi(arttInt); ++i) {
                arttOutput += std::string(arttXBound - arttCharsInLine, '.') + '\n';
                arttCharsInLine = 0;
            }
            arttInt.clear();
        } else if (c == '!') {
            arttOutput += std::string(arttXBound - arttCharsInLine, '.') + '\n';
            break;
        } else {
            char arttState = (c == 'b') ? '.' : '*';
            if (arttInt.empty()) arttInt = "1";
            arttOutput += std::string(std::stoi(arttInt), arttState);
            arttCharsInLine += std::stoi(arttInt);
            arttInt.clear();
        }
    }

    std::vector<std::string> arttOutputLines;
    std::smatch match;
    std::string::const_iterator searchStart(arttOutput.cbegin());
    while (std::regex_search(searchStart, arttOutput.cend(), match, line_regex)) {
        arttOutputLines.push_back(match[1]);
        searchStart = match.suffix().first;
    }

    assert(arttOutputLines.size() == arttYBound);
    for (const auto& line : arttOutputLines) {
        assert(line.size() == arttXBound);
    }

    return arttOutput.substr(0, arttOutput.size() - 1); // remove extra '\n' at the end of str
}

std::string matrix_to_txt(const std::unordered_map<std::pair<int, int>, std::vector<std::vector<bool>>>& mttGrid) {
    int mttUpMost = mttGrid.begin()->first.second;
    int mttDownMost = mttGrid.begin()->first.second;
    int mttLeftMost = mttGrid.begin()->first.first;
    int mttRightMost = mttGrid.begin()->first.first;

    for (const auto& chunk : mttGrid) {
        if (chunk.first.second > mttUpMost) mttUpMost = chunk.first.second;
        if (chunk.first.second < mttDownMost) mttDownMost = chunk.first.second;
        if (chunk.first.first > mttRightMost) mttRightMost = chunk.first.first;
        if (chunk.first.first < mttLeftMost) mttLeftMost = chunk.first.first;
    }

    int mttHeight = (mttUpMost - mttDownMost + 1) * 8;
    int mttWidth = (mttRightMost - mttLeftMost + 1) * 8;

    std::vector<std::vector<char>> mttOutput(mttWidth, std::vector<char>(mttHeight, '.'));

    for (const auto& chunk : mttGrid) {
        std::pair<int, int> mttZeroedChunk = {chunk.first.first - mttLeftMost, chunk.first.second - mttDownMost};
        for (int x = 0; x < 8; ++x) {
            for (int y = 0; y < 8; ++y) {
                if (chunk.second[x][y]) {
                    try {
                        mttOutput[mttZeroedChunk.first * 8 + x][mttZeroedChunk.second * 8 + y] = '*';
                    } catch (const std::out_of_range&) {
                        std::cerr << "IndexError. These coords out of range: " << chunk.first.first * 8 + x << " " << chunk.first.second * 8 + y << std::endl;
                    }
                }
            }
        }
    }

    std::string mttOutputStr;
    for (int y = mttOutput[0].size() - 1; y >= 0; --y) {
        for (int x = 0; x < mttOutput.size(); ++x) {
            mttOutputStr += mttOutput[x][y];
        }
        mttOutputStr += '\n';
    }

    return mttOutputStr.substr(0, mttOutputStr.size() - 1); // shave final '\n'
}

std::string txt_to_RLE(const std::string& ttrInput, const std::string& ttrComment) {
    std::string ttrInputWithNewline = ttrInput + '\n';
    std::vector<std::string> ttrLines;
    std::smatch match;
    std::string::const_iterator searchStart(ttrInputWithNewline.cbegin());
    while (std::regex_search(searchStart, ttrInputWithNewline.cend(), match, line_regex)) {
        ttrLines.push_back(match[1]);
        searchStart = match.suffix().first;
    }

    int ttrHeight = ttrLines.size();
    int ttrWidth = ttrLines[0].size();
    for (const auto& line : ttrLines) {
        assert(line.size() == ttrWidth);
    }

    std::string ttrOutputRLE;
    for (const auto& line : ttrLines) {
        std::pair<char, int> ttrRun = {line[0], 0};
        for (char cell : line) {
            if (cell == ttrRun.first) {
                ttrRun.second++;
            } else {
                char ttrState = (ttrRun.first == '.') ? 'b' : 'o';
                ttrOutputRLE += (ttrRun.second == 1) ? std::string(1, ttrState) : std::to_string(ttrRun.second) + ttrState;
                ttrRun = {cell, 1};
            }
        }
        if (ttrRun.first == '*') {
            ttrOutputRLE += (ttrRun.second == 1) ? "o" : std::to_string(ttrRun.second) + "o";
        }
        ttrOutputRLE += '$';
    }
    ttrOutputRLE.back() = '!';

    std::vector<std::string> ttrDollars;
    searchStart = ttrOutputRLE.cbegin();
    while (std::regex_search(searchStart, ttrOutputRLE.cend(), match, greedy_regex)) {
        ttrDollars.push_back(match[0]);
        searchStart = match.suffix().first;
    }

    int ttrDollarCount = 0;
    for (const auto& instance : ttrDollars) {
        if (instance.size() > ttrDollarCount) ttrDollarCount = instance.size();
    }

    for (int len = ttrDollarCount; len > 1; --len) {
        std::regex dollarSub(std::string(len, '$'));
        ttrOutputRLE = std::regex_replace(ttrOutputRLE, dollarSub, std::to_string(len) + '$');
    }

    std::string ttrOutputHeader = "x = " + std::to_string(ttrWidth) + ", y = " + std::to_string(ttrHeight) + ", rule = B3/S23\n";
    return ttrComment + ttrOutputHeader + ttrOutputRLE;
}

std::unordered_map<std::string, std::string> comment_to_dict(const std::string& ctdInput) {
    std::smatch match;
    std::regex_search(ctdInput, match, comment_regex);
    std::string ctdTheComment = match[0];
    std::vector<std::vector<std::string>> ctdCommentItems;
    searchStart = ctdTheComment.cbegin();
    while (std::regex_search(searchStart, ctdTheComment.cend(), match, comment_items_regex)) {
        ctdCommentItems.push_back({match[1], match[2], match[3]});
        searchStart = match.suffix().first;
    }

    std::vector<std::vector<std::string>> ctdList4Output;
    for (const auto& item : ctdCommentItems) {
        ctdList4Output.push_back({item[0]});
        if (!item[1].empty()) {
            ctdList4Output.back().push_back("");
            std::string ctdInt;
            for (char c : item[1]) {
                if (std::isdigit(c)) {
                    ctdInt += c;
                } else {
                    ctdList4Output.back().back() += ctdInt;
                    ctdInt.clear();
                }
            }
        } else {
            ctdList4Output.back().push_back(item[2]);
        }
    }

    if (ctdList4Output[0][0][0] == ' ') {
        ctdList4Output[0][0] = ctdList4Output[0][0].substr(1);
    }

    for (auto& item : ctdList4Output) {
        item[0] = std::regex_replace(item[0], std::regex(" "), "_");
        if (item[1].size() == 1) {
            item[1] = item[1][0];
        }
    }

    std::unordered_map<std::string, std::string> ctdOutput;
    for (const auto& item : ctdList4Output) {
        ctdOutput[item[0]] = item[1];
    }
    return ctdOutput;
}

std::unordered_map<std::pair<int, int>, std::vector<std::vector<bool>>> txt_to_matrix(const std::string& ttmGrid) {
    std::string ttmGridWithNewline = ttmGrid + '\n';
    std::vector<std::string> ttmGridLines;
    std::smatch match;
    std::string::const_iterator searchStart(ttmGridWithNewline.cbegin());
    while (std::regex_search(searchStart, ttmGridWithNewline.cend(), match, line_regex)) {
        ttmGridLines.push_back(match[1]);
        searchStart = match.suffix().first;
    }
    std::reverse(ttmGridLines.begin(), ttmGridLines.end());

    int ttmCellWidth = ttmGridLines[0].size();
    int ttmCellHeight = ttmGridLines.size();
    int ttmChunkWidth = (ttmCellWidth % 8 == 0) ? ttmCellWidth / 8 : ttmCellWidth / 8 + 1;
    int ttmChunkHeight = (ttmCellHeight % 8 == 0) ? ttmCellHeight / 8 : ttmCellHeight / 8 + 1;

    std::unordered_map<std::pair<int, int>, std::vector<std::vector<bool>>> ttmOutGrid;
    for (int chunkY = 0; chunkY < ttmChunkHeight; ++chunkY) {
        for (int chunkX = 0; chunkX < ttmChunkWidth; ++chunkX) {
            std::vector<std::vector<bool>> chunk(8, std::vector<bool>(8, false));
            for (int cellY = 0; cellY < 8; ++cellY) {
                for (int cellX = 0; cellX < 8; ++cellX) {
                    try {
                        chunk[cellX][cellY] = (ttmGridLines[chunkY * 8 + cellY][chunkX * 8 + cellX] == '*');
                    } catch (const std::out_of_range&) {
                        chunk[cellX][cellY] = false;
                    }
                }
            }
            if (chunk != empty_chunk) {
                ttmOutGrid[{chunkX, chunkY}] = chunk;
            }
        }
    }
    return ttmOutGrid;
}

std::unordered_map<std::pair<int, int>, std::vector<std::vector<bool>>> RLE_to_matrix(const std::string& rtmInput) {
    return txt_to_matrix(easy_RLE_to_txt(rtmInput));
}

std::string matrix_to_RLE(const std::unordered_map<std::pair<int, int>, std::vector<std::vector<bool>>>& mtrMatrix, const std::string& mtrComment) {
    return txt_to_RLE(matrix_to_txt(mtrMatrix), mtrComment);
}

void pro_print_grid(const std::unordered_map<std::pair<int, int>, std::vector<std::vector<bool>>>& ppgGrid, const std::pair<int, int>& ppgUpLeft, const std::pair<int, int>& ppgDownRight) {
    std::string ppgOutput;
    for (const auto& chunkRow : get_chunk_window(ppgUpLeft, ppgDownRight)) {
        for (int cellRow = 7; cellRow >= 0; --cellRow) {
            for (const auto& chunk : chunkRow) {
                for (int x = 0; x < 8; ++x) {
                    if (ppgGrid.find(chunk) != ppgGrid.end()) {
                        ppgOutput += (ppgGrid.at(chunk)[x][cellRow] ? "[]" : "<>");
                    } else {
                        ppgOutput += "::";
                    }
                }
            }
            ppgOutput += '\n';
        }
    }
    std::cout << ppgOutput;
}

std::string grid_to_string(const std::unordered_map<std::pair<int, int>, std::vector<std::vector<bool>>>& gtsGrid, const std::pair<int, int>& gtsUpLeft, const std::pair<int, int>& gtsDownRight) {
    std::string gtsOutput;
    for (const auto& chunkRow : get_chunk_window(gtsUpLeft, gtsDownRight)) {
        for (int cellRow = 7; cellRow >= 0; --cellRow) {
            for (const auto& chunk : chunkRow) {
                for (int x = 0; x < 8; ++x) {
                    if (gtsGrid.find(chunk) != gtsGrid.end()) {
                        gtsOutput += (gtsGrid.at(chunk)[x][cellRow] ? '1' : '0');
                    } else {
                        gtsOutput += '0';
                    }
                }
            }
            gtsOutput += '\n';
        }
    }
    return gtsOutput;
}

std::vector<std::vector<std::pair<int, int>>> get_chunk_window(const std::pair<int, int>& gcwUpLeft, const std::pair<int, int>& gcwDownRight) {
    assert(gcwUpLeft.first <= gcwDownRight.first && gcwUpLeft.second >= gcwDownRight.second);
    std::vector<std::vector<std::pair<int, int>>> gcwOutput;
    for (int y = gcwUpLeft.second; y >= gcwDownRight.second; --y) {
        std::vector<std::pair<int, int>> row;
        for (int x = gcwUpLeft.first; x <= gcwDownRight.first; ++x) {
            row.emplace_back(x, y);
        }
        gcwOutput.push_back(row);
    }
    assert(!gcwOutput.empty());
    return gcwOutput;
}

File: lib/convert_functions.py
#this file houses all of the functions and regexes and such

#TODO remove print lines that are commented out
#TODO check for consistency in cell values and such
#NOTE: prefix all the regex strings with r to avoid escape character issues.
#      idk why this happens but it does. I found a random stackoverflow post
#      that said to do it and it worked. I added it to all current regexes
#      just make sure when you add new ones you do it too.  Example: r'(\d+)'

#NOTE do not include a newline at the end of a custom CGOL comment

import re
from lib import regexes as r
# Functions that convert pattern types:
#This function converts RLE to plaintext. It accepts the entire unedited RLE string. It does not preserve comments.
def easy_RLE_to_txt(erttRLE):
    global RLE_regex, whitespace_regex
    erttRLEfound = re.search(r.RLE_regex, erttRLE)
    if erttRLEfound != None:
        erttGroups = erttRLEfound.group(3, 4, 5)
        #remove whitespace
        erttRLEnoWS = ''
        for char in erttGroups[2]:
            if not char in ('\n', '\t', ' '):
                erttRLEnoWS = erttRLEnoWS + char
        return advanced_RLE_to_txt(int(erttGroups[0]), int(erttGroups[1]), erttRLEnoWS)


#This function converts RLE to plaintext. It accepts the pattern RLE, the X bound of the pattern, and the Y bound.
def advanced_RLE_to_txt(arttXBound, arttYBound, arttRLE):
    global line_regex, digits
    arttRLE = arttRLE.lower()
    arttOutput = ''
    arttInt = ''
    arttCharsInLine = 0

    for char in arttRLE:
        if  char in ('\n', '\t', ' '):
            continue
        if char in r.digits:
            arttInt += str(char)

        elif char == '$':
            if arttInt == '': arttInt = '1'
            for arttLine in range(int(arttInt)):
                arttOutput += '.' * (arttXBound-arttCharsInLine) + '\n'
                arttCharsInLine = 0
                arttInt = ''

        elif char == '!':
            arttOutput += '.' * (arttXBound-arttCharsInLine) + '\n'
            break

        else:# o is live, b is dead. Anything that is not b will be treated as o.
            arttState = '.' if char == 'b' else '*'
            if arttInt == '': arttInt = '1'
            arttOutput += arttState * int(arttInt)
            arttCharsInLine += int(arttInt)
            arttInt = ''

    #ensure that output is not flawed via asserts
    arttOutputLines = r.line_regex.findall(arttOutput)
    assert len(arttOutputLines) == arttYBound, str(len(arttOutputLines)) + ' != ' + str(arttYBound)
    for arttLine in arttOutputLines: assert len(arttLine) == arttXBound, str(len(arttLine)) + ' != ' + str(arttXBound)
    return arttOutput[:-1] #remove extra '\n' at the end of str


#converts list matrix to plaintext
def matrix_to_txt(mttGrid):
    #find bounds of grid by checking all chunks and recording the furthest ones
    mttUpMost = tuple(mttGrid.keys())[0][1]
    mttDownMost = tuple(mttGrid.keys())[0][1]
    mttLeftMost = tuple(mttGrid.keys())[0][0]
    mttRightMost = tuple(mttGrid.keys())[0][0]
    for mttChunk in mttGrid:
        if mttChunk[1] > mttUpMost: mttUpMost = mttChunk[1]
        elif mttChunk[1] < mttDownMost: mttDownMost = mttChunk[1]
        if mttChunk[0] > mttRightMost: mttRightMost = mttChunk[0]
        elif mttChunk[0] < mttLeftMost: mttLeftMost = mttChunk[0]
    mttHeight = (mttUpMost-mttDownMost+1)*8
    mttWidth = (mttRightMost-mttLeftMost+1)*8

    #create empty plaintext grid
    mttOutput = []
    for mttRow in range(mttWidth): mttOutput.append(['.'] * mttHeight) # X and Y got switched?

    #fill in live cells
    for mttChunk in mttGrid:
        #compensates chunk positions in mttOutput.
        mttZeroedChunk = (mttChunk[0] - mttDownMost, mttChunk[1] - mttLeftMost)
        for mttX in range(8):
            for mttY in range(8):
                if mttGrid[mttChunk][mttX][mttY]:
                    try: #overshoots index by 1? #TODO remove try
                        mttOutput[mttZeroedChunk[0] * 8 + mttX][mttZeroedChunk[1] * 8 + mttY] = '*' #basically uses abs pos.
                    except IndexError:
                        print('IndexError. These coords out of range: %s %s' % (mttChunk[0] * 8 + mttX, mttChunk[1] * 8 + mttY))
                        print(mttChunk, mttZeroedChunk, mttX, mttY)

    #TODO shave empty edges

    #convert to string.
    mttOutputStr = ''
    for mttY in range(len(mttOutput[0])-1, -1, -1): #inverts Y
        for mttX in range(len(mttOutput)):
            mttOutputStr += mttOutput[mttX][mttY]
        mttOutputStr += '\n'

    return mttOutputStr[:-1] # shave final '\n'


#converts txt to RLE and provides meta-data comments:
def txt_to_RLE(ttrInput, ttrComment=r.default_RLE_comment): #TODO accept comment to use in RLE instead of default.
    global line_regex, greedy_regex
    ttrInput += '\n'
    ttrLines = r.line_regex.findall(ttrInput)
    #get dimensions of txt
    ttrHeight = len(ttrLines)
    ttrWidth = len(ttrLines[0])
    #check if each line is equal in length
    for ttrLine in ttrLines: assert len(ttrLine) == ttrWidth, str(len(ttrLine)) + ' != ' + str(ttrWidth)

    ttrOutputRLE = ''
    for ttrLine in ttrLines:
        #ttrOutputRLE.append('')
        ttrRun = [ttrLine[0], 0]
        for ttrCell in ttrLine:
            if ttrCell == ttrRun[0]:
                ttrRun[1] += 1 #counts how many conxecutive chars are present
            else:#TODO fix for when ttrRun[0] == ''# probably unecessary now that line is: ttrRun = [ttrLine[0], 0]
                ttrState = 'b' if ttrRun[0] == '.' else 'o'
                ttrRunCount = '' if ttrRun[1] == 1 else str(ttrRun[1])
                ttrOutputRLE += ttrRunCount + ttrState
                ttrRun = [ttrCell, 1]
        if ttrRun[0] == '*':
            ttrRunCount = '' if ttrRun[1] == 1 else str(ttrRun[1])
            ttrOutputRLE += ttrRunCount + 'o'
        ttrOutputRLE += '$'
    ttrOutputRLE = ttrOutputRLE[:-1] + '!' #replace final char with ! instead of $

    #search for clusters of $ (e.g. '...$$$...') and convert to '...3$...'
    ttrDollars = r.greedy_regex.findall(ttrOutputRLE)
    ttrDollarCount = 0
    for ttrInstance in ttrDollars: #finds the longest instance
        if len(ttrInstance) > ttrDollarCount: ttrDollarCount = len(ttrInstance)
    for ttrDollarLen in range(ttrDollarCount, 1, -1): #uses .sub on all $clusters, starting with the longest ones
        ttrDollarSub = re.compile('\\$' * ttrDollarLen)
        ttrOutputRLE = ttrDollarSub.sub('%s$' % ttrDollarLen, ttrOutputRLE)

    ttrOutputHeader = 'x = %s, y = %s, rule = B3/S23\n' % (ttrWidth, ttrHeight)
    return ttrComment + ttrOutputHeader + ttrOutputRLE


#accepts entire raw RLE, returns a dict of meta-data from the "#C [[ ZOOM 7 ]]" comment
def comment_to_dict(ctdInput):
    global comment_regex, comment_items_regex, digits
    ctdTheComment = r.comment_regex.search(ctdInput).group()
    ctdCommentItems = r.comment_items_regex.findall(ctdTheComment)

    #clean up keys and values
    ctdList4Output = []
    for ctdItem in ctdCommentItems:
        ctdList4Output.append([ctdItem[0]]) #deepcopy needed?
        if ctdItem[1] != '':
            ctdList4Output[-1].append([])
            ctdInt = ''
            for ctdChar in ctdItem[1]:
                if ctdChar in r.digits:
                    ctdInt += ctdChar
                else:
                    ctdList4Output[-1][1].append(int(ctdInt))
                    ctdInt = ''
        else:
            ctdList4Output.append([ctdItem[0], ctdItem[2]])

    if ctdList4Output[0][0][0] == ' ': #clip off leading space of first key
        ctdList4Output[0][0] = ctdList4Output[0][0][1:]
    #Lowercase the keys
    for ctdItem in ctdList4Output:
        ctdItem[0] = ctdItem[0].lower()
        ctdItem[0] = ctdItem[0].replace(' ', '_') #remove spaces from values
        if len(ctdItem[1]) == 1:        #convert single item lists to items eg. numbers
            ctdItem[1] = ctdItem[1][0]

    ctdOutput = {}
    for ctdItem in ctdList4Output:
        ctdOutput[ctdItem[0]] = ctdItem[1]
    return ctdOutput


#Converts plaintext pattern to list matrix pattern
def txt_to_matrix(ttmGrid):
    global line_regex, empty_chunk
    ttmGrid += '\n'#regex will not catch final line if it does not end with '\n'

    #break up grid into lines
    ttmGridLines = r.line_regex.findall(ttmGrid)
    ttmGridLines.reverse()

    #get grid specs
    ttmCellWidth = len(ttmGridLines[0]) #If this line raises an error, it is because cfGridLines == []. cfGrid is invalid and likely missing \t or \n
    ttmCellHeight = len(ttmGridLines)
    if ttmCellWidth % 8 == 0:
        ttmChunkWidth = ttmCellWidth // 8
    else:
        ttmChunkWidth = ttmCellWidth // 8 + 1
    if ttmCellHeight % 8 == 0:
        ttmChunkHeight = ttmCellHeight // 8
    else:
        ttmChunkHeight = ttmCellHeight // 8 + 1

    #convert to dictionary
    ttmOutGrid = {}
    for ChunkY in range(ttmChunkHeight):
        for ChunkX in range(ttmChunkWidth):
            ttmOutGrid[(ChunkX, ChunkY)] = [[None for _ in range(8)] for _ in range(8)] #Suggested by Copilot
            #ttmOutGrid[(ChunkX, ChunkY)] = copy.deepcopy([[None]*8]*8) #my design. reference glitches?
            for CellY in range(8):
                for CellX in range(8):
                    try:
                        #                                                       7- is used because Y0 is at the bottom of the txt
                        #ttmOutGrid[(ChunkX, ChunkY)][CellX][CellY] = (True if ttmGridLines[7 - (ChunkY*8+CellY)][ChunkX*8+CellX]=='*' else False) #old code. fixed?
                        ttmOutGrid[(ChunkX, ChunkY)][CellX][CellY] = (True if ttmGridLines[ChunkY*8+CellY][ChunkX*8+CellX] == '*' else False)
                    except IndexError:
                        ttmOutGrid[(ChunkX, ChunkY)][CellX][CellY] = False
            #delete chunk if empty
            if ttmOutGrid[(ChunkX, ChunkY)] == r.empty_chunk:
                del ttmOutGrid[(ChunkX, ChunkY)]
    return ttmOutGrid


#Converts RLE to list matrix pattern
def RLE_to_matrix(rtmInput):
    return txt_to_matrix(easy_RLE_to_txt(rtmInput))

#Converts list-matrix to RLE and adds default comment line
def matrix_to_RLE(mtrMatrix, mtrComment=r.default_RLE_comment):
    return txt_to_RLE(matrix_to_txt(mtrMatrix), ttrComment=mtrComment)


#Other functions:


#Permanantly borrowed™ from CGOL_game_runner
#accepts grid and window for camera, prints grid to terminal.
def pro_print_grid(ppgGrid, ppgUpLeft, ppgDownRight):
    ppgOutput = ''
    for ppgChunkRow in get_chunk_window(ppgUpLeft, ppgDownRight):
        for ppgCellRow in range(7, -1, -1):
            for ppgChunk in ppgChunkRow:
                for ppgX in range(8):
                    if ppgChunk in ppgGrid:
                        ppgOutput = ppgOutput + ('[]' if ppgGrid[ppgChunk][ppgX][ppgCellRow] else '<>')
                    else:
                        ppgOutput += '::'

            ppgOutput += '\n'
    print(ppgOutput)


#This is sort of a combo of matrix_to_txt and pro_print_grid. It accepts a list-matrix, the coord of a chunk (camera window position), and
#returns a 64x64 plaintext grid.
def grid_to_string(gtsGrid, gtsUpLeft, gtsDownRight):
    gtsOutput = ''
    for gtsChunkRow in get_chunk_window(gtsUpLeft, gtsDownRight):
        for gtsCellRow in range(7, -1, -1):
            for gtsChunk in gtsChunkRow:
                for gtsX in range(8):
                    if gtsChunk in gtsGrid:
                        gtsOutput = gtsOutput + ('1' if gtsGrid[gtsChunk][gtsX][gtsCellRow] else '0') # Loaded chunks
                    else:
                        gtsOutput += '0' # Unloaded chunk

            gtsOutput += '\n'
    return gtsOutput

# WIP CENTERED THING THAT WON"T WORK
# def grid_to_string_centered(grid, coordinate, image_size=64):
#     center_chunk = (coordinate[0] // 8, coordinate[1] // 8)
#     center_chunk_offset = (coordinate[0] % 8, coordinate[1] % 8)
#     chunk_count = (image_size // 8) + 2
#     top_left_chunk = ((coordinate[0] // 8) - (chunk_count // 2), (coordinate[1] // 8) + (chunk_count // 2))
#     bottom_right_chunk = ((coordinate[0] // 8) + (chunk_count // 2), (coordinate[1] // 8) - (chunk_count // 2))
#     print(top_left_chunk, bottom_right_chunk)
#     gtsOutput = ''
#     for gtsChunkRow in range(top_left_chunk[1], bottom_right_chunk[1]-1, -1):
#         for gtsCellRow in range(7, -1, -1):
#             for gtsChunk in range(top_left_chunk[0], bottom_right_chunk[0]):
#                 for gtsX in range(8):
#                     if (gtsChunk, gtsChunkRow) in grid:
#                         gtsOutput = gtsOutput + ('1' if grid[(gtsChunk, gtsChunkRow)][gtsX][gtsCellRow] else '0') # Loaded chunks
#                     else:
#                         gtsOutput += '0' # Unloaded chunk

#             gtsOutput += '\n'
#     print(gtsOutput)
#     return gtsOutput


#accepts coordinates of two chunks and returns all chunks within the window.
def get_chunk_window(gcwUpLeft, gcwDownRight):
    assert type(gcwUpLeft) in (tuple, list) and len(gcwUpLeft) == 2, 'gcw parameter gcwUpLeft passed invalid argument' #TODO remove asserts for efficiency
    assert type(gcwDownRight) in (tuple, list) and len(gcwDownRight) == 2, 'gcw parameter gcwDownRight passed invalid argument'
    gcwOutput = []
    for gcwY, gcwCounter in zip(range(gcwUpLeft[1], gcwDownRight[1]-1, -1), range(99)): #idk why range(99) is used. shrug.
        gcwOutput.append([])
        for gcwX in range(gcwUpLeft[0], gcwDownRight[0]+1):
            gcwOutput[gcwCounter].append((gcwX, gcwY))
    assert gcwOutput != []
    return gcwOutput
